packages/hint-doctype/src/hint.ts

```diff
@@ -0,0 +1,125 @@
+ */
+
+import { Category } from 'hint/dist/src/lib/enums/category';
+import { HintScope } from 'hint/dist/src/lib/enums/hintscope';
+import { HintContext } from 'hint/dist/src/lib/hint-context';
+import { IHint, HintMetadata, FetchEnd, IAsyncHTMLElement } from 'hint/dist/src/lib/types';
+import { debug as d } from 'hint/dist/src/lib/utils/debug';
+
+const debug: debug.IDebugger = d(__filename);
+
+/*
+ * ------------------------------------------------------------------------------
+ * Public
+ * ------------------------------------------------------------------------------
+ */
+
+export default class implements IHint {
+
+    public static readonly meta: HintMetadata = {
+        docs: {
+            category: Category.interoperability,
+            description: `Hint to validate if the doctype is correct`
+        },
+        id: 'doctype',
+        schema: [],
+        scope: HintScope.any
+    }
+
+    public constructor(context: HintContext) {
+
+        const doctypeRegexFactory = (flags?: string) => {
+            return new RegExp(/(<!doctype\s(html)\s*?)(>)(.+)?/, flags ? flags : 'g');
+        };
+
+        const checkDoctypeIsValid = async (resource: string, element: IAsyncHTMLElement | null, content: string): Promise<boolean> => {
+            debug(`Checking if the doctype is valid.`);
+
+            const matched = content.match(doctypeRegexFactory('gi'));
+
+            if (!matched || matched.length < 1) {
+                await context.report(resource, element, `The file does not contain a valid doctype tag`);
+
+                return false;
+            }
+
+            return true;
+        };
+
+        const checkDoctypeFirstLine = async (resource: string, element: IAsyncHTMLElement | null, content: string): Promise<void> => {
+            debug(`Checking if the doctype is in the first line.`);
+
+            const firstLine = content.split(/\r|\n/)[0];
+            const matched = firstLine.match(doctypeRegexFactory('gi'));
+
+
+            if (!matched || matched.length < 1) {
+                await context.report(resource, element, `The first line does not contain a valid doctype tag.`);
+
+                return;
+            }
+
+            if (matched) {
+                const cleaned = matched[0].trim().split('');
+
+                if (cleaned[cleaned.length-1] !== '>') {
+                    await context.report(resource, element, `There is additional information on the line with the doctype tag`);
+
+                    return;
+                }
+            }
+        };
+
+        const checkDoctypeLowercase = async (resource: string, element: IAsyncHTMLElement | null, content: string): Promise<void> => {
+            debug(`Checking that the doctype is in lowercase`);
+
+            const matched = content.match(doctypeRegexFactory());
+
+            if (!matched) {
+                await context.report(resource, element, `The doctype should be in lowercase`);
```

Comment:
> DOCTYPE should be lowercase.

Also we should talk about enforcing this or not. The reason why we say lowercase is for (marginally) better compression but even [the spec uses `<!DOCTYPE html>`](https://html.spec.whatwg.org/multipage/syntax.html#the-doctype). I believe this is going to cause more noise than anything else so maybe we should just ignore the casing? @webhintio/contributors thoughts?

Also the spec mentions another valid DOCTYPE: `<!DOCTYPE html SYSTEM "about:legacy-compat">`. That I don't believe it's covered in here or the tests.

Reply:
I agree we should ignore the casing. Although webhint is configurable, we should be careful about flagging issues where there's limited benefit in a change. Developers should be able to trust that webhint identifies *important* issues in their code.

Reply:
The documentation needs to be updated too.